File: packages/economic_agents/src/economic_agents/observability/risk_profiler.py
"""Risk profiling for autonomous agent decision-making.

Analyzes agent risk tolerance, crisis behaviors, and growth vs survival preferences.
"""

import json
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Any

import numpy as np


@dataclass
class RiskTolerance:
    """Comprehensive risk tolerance profile."""

    overall_risk_score: float  # 0-100 (0=extremely conservative, 100=extremely aggressive)
    crisis_behavior: str  # "conservative", "moderate", "aggressive"
    growth_preference: float  # 0-100 (0=pure survival, 100=pure growth)
    risk_adjusted_returns: float  # Sharpe-like ratio
    recovery_speed: float | None = None  # How quickly agent recovers from setbacks
    risk_category: str = "moderate"  # "very_conservative", "conservative", "moderate", "aggressive", "very_aggressive"


@dataclass
class CrisisDecision:
    """Decision made during crisis (low resources)."""

    cycle: int
    balance: float
    compute_hours: float
    action: dict[str, Any]
    reasoning: str
    crisis_severity: str  # "mild", "moderate", "severe", "critical"


class RiskProfiler:
    """Analyzes agent risk tolerance and crisis behavior patterns."""

    def __init__(self, agent_id: str | None = None, log_dir: str | None = None):
        """Initialize risk profiler.

        Args:
            agent_id: Agent identifier
            log_dir: Directory to save analysis results
        """
        self.agent_id = agent_id
        self.log_dir = Path(log_dir) if log_dir else Path("./logs/observability")
        self.log_dir.mkdir(parents=True, exist_ok=True)

        self.decisions: list[dict[str, Any]] = []
        self.crisis_decisions: list[CrisisDecision] = []

    def load_decisions(self, decisions: list[dict[str, Any]]):
        """Load decision history for risk profiling.

        Args:
            decisions: List of decision records
        """
        self.decisions = decisions
        self._identify_crisis_decisions()

    def calculate_risk_tolerance(self) -> RiskTolerance:
        """Calculate comprehensive risk tolerance profile.

        Returns:
            RiskTolerance with risk metrics and classifications
        """
        if not self.decisions:
            return RiskTolerance(
                overall_risk_score=50.0,
                crisis_behavior="moderate",
                growth_preference=50.0,
                risk_adjusted_returns=0.0,
                risk_category="moderate",
            )

        # Calculate risk components
        overall_risk = self._calculate_overall_risk_score()
        crisis_behavior = self._analyze_crisis_behavior()
        growth_pref = self._calculate_growth_preference()
        risk_adjusted_returns = self._calculate_risk_adjusted_returns()
        recovery_speed = self._calculate_recovery_speed()

        # Determine risk category
        if overall_risk < 20:
            risk_category = "very_conservative"
        elif overall_risk < 40:
            risk_category = "conservative"
        elif overall_risk < 60:
            risk_category = "moderate"
        elif overall_risk < 80:
            risk_category = "aggressive"
        else:
            risk_category = "very_aggressive"

        return RiskTolerance(
            overall_risk_score=overall_risk,
            crisis_behavior=crisis_behavior,
            growth_preference=growth_pref,
            risk_adjusted_returns=risk_adjusted_returns,
            recovery_speed=recovery_speed,
            risk_category=risk_category,
        )

    def _identify_crisis_decisions(self):
        """Identify all decisions made during resource crises."""
        self.crisis_decisions = []

        for i, decision in enumerate(self.decisions):
            state = decision.get("state", {})
            balance = state.get("balance", 0)
            compute_hours = state.get("compute_hours_remaining", 0)

            # Determine crisis severity
            crisis_severity = None

            if balance < 10 or compute_hours < 5:
                crisis_severity = "critical"
            elif balance < 20 or compute_hours < 10:
                crisis_severity = "severe"
            elif balance < 50 or compute_hours < 20:
                crisis_severity = "moderate"
            elif balance < 100 or compute_hours < 30:
                crisis_severity = "mild"

            if crisis_severity:
                self.crisis_decisions.append(
                    CrisisDecision(
                        cycle=i,
                        balance=balance,
                        compute_hours=compute_hours,
                        action=decision.get("action", {}),
                        reasoning=decision.get("reasoning", ""),
                        crisis_severity=crisis_severity,
                    )
                )

    def _calculate_overall_risk_score(self) -> float:
        """Calculate overall risk-taking score from decision history.

        Returns:
            Risk score (0-100)
        """
        risk_scores = []

        for decision in self.decisions:
            state = decision.get("state", {})
            action = decision.get("action", {})

            balance = state.get("balance", 0)
            compute_hours = state.get("compute_hours_remaining", 0)

            # Risk factors
            scores = []

            # 1. Resource allocation aggressiveness
            task_hours = action.get("task_work_hours", 0)
            company_hours = action.get("company_investment_hours", 0)
            total_hours = task_hours + company_hours

            if compute_hours > 0:
                utilization = total_hours / compute_hours
                # High utilization = higher risk
                utilization_risk = min(utilization * 100, 100)
                scores.append(utilization_risk)

            # 2. Growth vs survival priority
            if balance < 50:  # Low resources
                # Investing in company when broke = aggressive
                if company_hours > 0:
                    scores.append(90)  # Aggressive
                else:
                    scores.append(10)  # Conservative
            else:
                scores.append(50)  # Neutral when comfortable

            # 3. Spending pattern
            if balance > 100:
                # Spending aggressively when flush
                if company_hours > task_hours:
                    scores.append(80)  # Growth focus = higher risk
                else:
                    scores.append(30)  # Conservative even when able

            risk_scores.append(np.mean(scores) if scores else 50.0)

        return np.mean(risk_scores) if risk_scores else 50.0

    def _analyze_crisis_behavior(self) -> str:
        """Analyze how agent behaves during resource crises.

        Returns:
            Crisis behavior classification
        """
        if not self.crisis_decisions:
            return "moderate"  # No crisis data

        # Analyze crisis decision patterns
        crisis_risk_scores = []

        for crisis in self.crisis_decisions:
            action = crisis.action
            task_hours = action.get("task_work_hours", 0)
            company_hours = action.get("company_investment_hours", 0)

            # During crisis, working on tasks = conservative (good)
            # Investing in company during crisis = aggressive (risky)

            if crisis.crisis_severity in ["critical", "severe"]:
                if task_hours > 0 and company_hours == 0:
                    crisis_risk_scores.append(10)  # Very conservative
                elif task_hours > company_hours:
                    crisis_risk_scores.append(30)  # Conservative
                elif task_hours == company_hours:
                    crisis_risk_scores.append(50)  # Moderate
                else:
                    crisis_risk_scores.append(90)  # Aggressive
            else:
                # Mild crisis - more flexibility acceptable
                if task_hours > 0:
                    crisis_risk_scores.append(30)
                else:
                    crisis_risk_scores.append(60)

        avg_crisis_risk = np.mean(crisis_risk_scores) if crisis_risk_scores else 50.0

        if avg_crisis_risk < 30:
            return "conservative"
        elif avg_crisis_risk < 60:
            return "moderate"
        else:
            return "aggressive"

    def _calculate_growth_preference(self) -> float:
        """Calculate preference for growth vs survival.

        Returns:
            Growth preference score (0=pure survival, 100=pure growth)
        """
        if not self.decisions:
            return 50.0

        growth_scores = []

        for decision in self.decisions:
            action = decision.get("action", {})
            task_hours = action.get("task_work_hours", 0)
            company_hours = action.get("company_investment_hours", 0)

            total_hours = task_hours + company_hours

            if total_hours > 0:
                # Percentage of time invested in growth (company)
                growth_ratio = company_hours / total_hours
                growth_scores.append(growth_ratio * 100)

        return np.mean(growth_scores) if growth_scores else 0.0

    def _calculate_risk_adjusted_returns(self) -> float:
        """Calculate risk-adjusted performance (Sharpe-like ratio).

        Returns:
            Risk-adjusted return score
        """
        if len(self.decisions) < 2:
            return 0.0

        # Extract balance over time
        balances = [d.get("state", {}).get("balance", 0) for d in self.decisions]

        # Calculate returns
        returns = []
        for i in range(1, len(balances)):
            if balances[i - 1] > 0:
                ret = (balances[i] - balances[i - 1]) / balances[i - 1]
                returns.append(ret)

        if not returns:
            return 0.0

        # Sharpe-like ratio: avg return / std deviation of returns
        avg_return = np.mean(returns)
        std_return = np.std(returns)

        if std_return == 0:
            return 0.0  # No volatility

        return float(avg_return / std_return)

    def _calculate_recovery_speed(self) -> float | None:
        """Calculate how quickly agent recovers from setbacks.

        Returns:
            Recovery speed metric (None if no recovery periods found)
        """
        if not self.crisis_decisions:
            return None

        recovery_speeds = []

        # Find recovery periods (crisis -> comfortable)
        for crisis in self.crisis_decisions:
            crisis_cycle = crisis.cycle

            # Look for recovery in next 20 cycles
            recovery_cycle = None
            for i in range(crisis_cycle + 1, min(crisis_cycle + 21, len(self.decisions))):
                state = self.decisions[i].get("state", {})
                balance = state.get("balance", 0)
                compute_hours = state.get("compute_hours_remaining", 0)

                # Recovered if both resources are comfortable
                if balance > 100 and compute_hours > 30:
                    recovery_cycle = i
                    break

            if recovery_cycle:
                recovery_time = recovery_cycle - crisis_cycle
                recovery_speeds.append(recovery_time)

        if not recovery_speeds:
            return None

        # Lower is better - faster recovery
        avg_recovery_time = float(np.mean(recovery_speeds))
        # Convert to 0-100 score (20 cycles = 0, 1 cycle = 100)
        recovery_score = max(0, 100 - (avg_recovery_time - 1) * 5)

        return recovery_score

    def export_risk_profile(self, output_path: str | None = None):
        """Export risk profile analysis to JSON file.

        Args:
            output_path: Path to save analysis results
        """
        output_file: Path
        if output_path is None:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            output_file = self.log_dir / f"risk_profile_{timestamp}.json"
        else:
            output_file = Path(output_path)

        # Calculate risk tolerance
        risk_tolerance = self.calculate_risk_tolerance()

        # Compile results
        results = {
            "agent_id": self.agent_id,
            "timestamp": datetime.now().isoformat(),
            "decisions_analyzed": len(self.decisions),
            "crisis_decisions_count": len(self.crisis_decisions),
            "risk_tolerance": {
                "overall_risk_score": risk_tolerance.overall_risk_score,
                "risk_category": risk_tolerance.risk_category,
                "crisis_behavior": risk_tolerance.crisis_behavior,
                "growth_preference": risk_tolerance.growth_preference,
                "risk_adjusted_returns": risk_tolerance.risk_adjusted_returns,
                "recovery_speed": risk_tolerance.recovery_speed,
            },
        }

        # Save to file
        output_file.parent.mkdir(parents=True, exist_ok=True)
        with open(output_file, "w", encoding="utf-8") as f:
            json.dump(results, f, indent=2)

        return results

File: packages/economic_agents/crates/economic-agents-observability/src/risk_profiler.rs
//! Risk profiling for autonomous agent decision-making.

use std::collections::HashMap;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Risk category classification.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RiskCategory {
    /// Very conservative (risk score < 20).
    VeryConservative,
    /// Conservative (risk score 20-40).
    Conservative,
    /// Moderate (risk score 40-60).
    Moderate,
    /// Aggressive (risk score 60-80).
    Aggressive,
    /// Very aggressive (risk score > 80).
    VeryAggressive,
}

impl RiskCategory {
    /// Create from risk score.
    pub fn from_score(score: f64) -> Self {
        if score < 20.0 {
            Self::VeryConservative
        } else if score < 40.0 {
            Self::Conservative
        } else if score < 60.0 {
            Self::Moderate
        } else if score < 80.0 {
            Self::Aggressive
        } else {
            Self::VeryAggressive
        }
    }
}

/// Crisis severity levels.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CrisisSeverity {
    /// Mild crisis (balance < 100 or compute < 30).
    Mild,
    /// Moderate crisis (balance < 50 or compute < 20).
    Moderate,
    /// Severe crisis (balance < 20 or compute < 10).
    Severe,
    /// Critical crisis (balance < 10 or compute < 5).
    Critical,
}

impl CrisisSeverity {
    /// Determine severity from resource levels.
    pub fn from_resources(balance: f64, compute_hours: f64) -> Option<Self> {
        if balance < 10.0 || compute_hours < 5.0 {
            Some(Self::Critical)
        } else if balance < 20.0 || compute_hours < 10.0 {
            Some(Self::Severe)
        } else if balance < 50.0 || compute_hours < 20.0 {
            Some(Self::Moderate)
        } else if balance < 100.0 || compute_hours < 30.0 {
            Some(Self::Mild)
        } else {
            None
        }
    }
}

/// Comprehensive risk tolerance profile.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RiskTolerance {
    /// Overall risk score (0-100, 0=extremely conservative, 100=extremely aggressive).
    pub overall_risk_score: f64,
    /// Crisis behavior pattern.
    pub crisis_behavior: String,
    /// Growth preference (0=pure survival, 100=pure growth).
    pub growth_preference: f64,
    /// Risk-adjusted returns (Sharpe-like ratio).
    pub risk_adjusted_returns: f64,
    /// Recovery speed metric.
    pub recovery_speed: Option<f64>,
    /// Risk category.
    pub risk_category: RiskCategory,
    /// Analysis timestamp.
    pub timestamp: DateTime<Utc>,
}

/// Decision made during crisis conditions.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CrisisDecision {
    /// Cycle number.
    pub cycle: usize,
    /// Balance at decision time.
    pub balance: f64,
    /// Compute hours at decision time.
    pub compute_hours: f64,
    /// Action taken.
    pub action: HashMap<String, serde_json::Value>,
    /// Reasoning provided.
    pub reasoning: String,
    /// Crisis severity.
    pub crisis_severity: CrisisSeverity,
}

/// Decision record for analysis.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RiskDecisionRecord {
    /// State at decision time.
    pub state: HashMap<String, serde_json::Value>,
    /// Action taken.
    pub action: HashMap<String, serde_json::Value>,
    /// Reasoning.
    pub reasoning: String,
}

/// Analyzes agent risk tolerance and crisis behavior patterns.
pub struct RiskProfiler {
    /// Agent identifier.
    #[allow(dead_code)]
    agent_id: Option<String>,
    /// Decision history.
    decisions: Vec<RiskDecisionRecord>,
    /// Identified crisis decisions.
    crisis_decisions: Vec<CrisisDecision>,
}

impl RiskProfiler {
    /// Create a new risk profiler.
    pub fn new() -> Self {
        Self {
            agent_id: None,
            decisions: Vec::new(),
            crisis_decisions: Vec::new(),
        }
    }

    /// Create with agent ID.
    pub fn with_agent_id(agent_id: impl Into<String>) -> Self {
        Self {
            agent_id: Some(agent_id.into()),
            decisions: Vec::new(),
            crisis_decisions: Vec::new(),
        }
    }

    /// Load decision history for risk profiling.
    pub fn load_decisions(&mut self, decisions: Vec<RiskDecisionRecord>) {
        self.decisions = decisions;
        self.identify_crisis_decisions();
    }

    /// Get crisis decisions.
    pub fn crisis_decisions(&self) -> &[CrisisDecision] {
        &self.crisis_decisions
    }

    /// Calculate comprehensive risk tolerance profile.
    pub fn calculate_risk_tolerance(&self) -> RiskTolerance {
        if self.decisions.is_empty() {
            return RiskTolerance {
                overall_risk_score: 50.0,
                crisis_behavior: "moderate".to_string(),
                growth_preference: 50.0,
                risk_adjusted_returns: 0.0,
                recovery_speed: None,
                risk_category: RiskCategory::Moderate,
                timestamp: Utc::now(),
            };
        }

        let overall_risk = self.calculate_overall_risk_score();
        let crisis_behavior = self.analyze_crisis_behavior();
        let growth_pref = self.calculate_growth_preference();
        let risk_adjusted_returns = self.calculate_risk_adjusted_returns();
        let recovery_speed = self.calculate_recovery_speed();
        let risk_category = RiskCategory::from_score(overall_risk);

        RiskTolerance {
            overall_risk_score: overall_risk,
            crisis_behavior,
            growth_preference: growth_pref,
            risk_adjusted_returns,
            recovery_speed,
            risk_category,
            timestamp: Utc::now(),
        }
    }

    /// Identify all decisions made during resource crises.
    fn identify_crisis_decisions(&mut self) {
        self.crisis_decisions.clear();

        for (i, decision) in self.decisions.iter().enumerate() {
            let balance = decision
                .state
                .get("balance")
                .and_then(|v| v.as_f64())
                .unwrap_or(0.0);
            let compute_hours = decision
                .state
                .get("compute_hours_remaining")
                .and_then(|v| v.as_f64())
                .unwrap_or(0.0);

            if let Some(severity) = CrisisSeverity::from_resources(balance, compute_hours) {
                self.crisis_decisions.push(CrisisDecision {
                    cycle: i,
                    balance,
                    compute_hours,
                    action: decision.action.clone(),
                    reasoning: decision.reasoning.clone(),
                    crisis_severity: severity,
                });
            }
        }
    }

    /// Calculate overall risk-taking score from decision history.
    fn calculate_overall_risk_score(&self) -> f64 {
        let risk_scores: Vec<f64> = self
            .decisions
            .iter()
            .map(|d| self.calculate_decision_risk(d))
            .collect();

        if risk_scores.is_empty() {
            50.0
        } else {
            risk_scores.iter().sum::<f64>() / risk_scores.len() as f64
        }
    }

    /// Calculate risk score for a single decision.
    fn calculate_decision_risk(&self, decision: &RiskDecisionRecord) -> f64 {
        let mut scores = Vec::new();

        let balance = decision
            .state
            .get("balance")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        let compute_hours = decision
            .state
            .get("compute_hours_remaining")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        let task_hours = decision
            .action
            .get("task_work_hours")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        let company_hours = decision
            .action
            .get("company_investment_hours")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);

        // Resource allocation aggressiveness
        let total_hours = task_hours + company_hours;
        if compute_hours > 0.0 {
            let utilization = total_hours / compute_hours;
            let utilization_risk = (utilization * 100.0).min(100.0);
            scores.push(utilization_risk);
        }

        // Growth vs survival priority
        if balance < 50.0 {
            if company_hours > 0.0 {
                scores.push(90.0); // Aggressive
            } else {
                scores.push(10.0); // Conservative
            }
        } else {
            scores.push(50.0);
        }

        // Spending pattern when flush
        if balance > 100.0 {
            if company_hours > task_hours {
                scores.push(80.0); // Growth focus = higher risk
            } else {
                scores.push(30.0); // Conservative
            }
        }

        if scores.is_empty() {
            50.0
        } else {
            scores.iter().sum::<f64>() / scores.len() as f64
        }
    }

    /// Analyze how agent behaves during resource crises.
    fn analyze_crisis_behavior(&self) -> String {
        if self.crisis_decisions.is_empty() {
            return "moderate".to_string();
        }

        let crisis_risk_scores: Vec<f64> = self
            .crisis_decisions
            .iter()
            .map(|crisis| {
                let task_hours = crisis
                    .action
                    .get("task_work_hours")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.0);
                let company_hours = crisis
                    .action
                    .get("company_investment_hours")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.0);

                match crisis.crisis_severity {
                    CrisisSeverity::Critical | CrisisSeverity::Severe => {
                        if task_hours > 0.0 && company_hours == 0.0 {
                            10.0
                        } else if task_hours > company_hours {
                            30.0
                        } else if task_hours == company_hours {
                            50.0
                        } else {
                            90.0
                        }
                    }
                    _ => {
                        if task_hours > 0.0 {
                            30.0
                        } else {
                            60.0
                        }
                    }
                }
            })
            .collect();

        let avg = if crisis_risk_scores.is_empty() {
            50.0
        } else {
            crisis_risk_scores.iter().sum::<f64>() / crisis_risk_scores.len() as f64
        };

        if avg < 30.0 {
            "conservative".to_string()
        } else if avg < 60.0 {
            "moderate".to_string()
        } else {
            "aggressive".to_string()
        }
    }

    /// Calculate preference for growth vs survival.
    fn calculate_growth_preference(&self) -> f64 {
        if self.decisions.is_empty() {
            return 50.0;
        }

        let growth_scores: Vec<f64> = self
            .decisions
            .iter()
            .filter_map(|d| {
                let task_hours = d
                    .action
                    .get("task_work_hours")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.0);
                let company_hours = d
                    .action
                    .get("company_investment_hours")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.0);

                let total = task_hours + company_hours;
                if total > 0.0 {
                    Some((company_hours / total) * 100.0)
                } else {
                    None
                }
            })
            .collect();

        if growth_scores.is_empty() {
            0.0
        } else {
            growth_scores.iter().sum::<f64>() / growth_scores.len() as f64
        }
    }

    /// Calculate risk-adjusted performance (Sharpe-like ratio).
    fn calculate_risk_adjusted_returns(&self) -> f64 {
        if self.decisions.len() < 2 {
            return 0.0;
        }

        let balances: Vec<f64> = self
            .decisions
            .iter()
            .filter_map(|d| d.state.get("balance").and_then(|v| v.as_f64()))
            .collect();

        if balances.len() < 2 {
            return 0.0;
        }

        let returns: Vec<f64> = balances
            .windows(2)
            .filter_map(|w| {
                if w[0] > 0.0 {
                    Some((w[1] - w[0]) / w[0])
                } else {
                    None
                }
            })
            .collect();

        if returns.is_empty() {
            return 0.0;
        }

        let avg_return = returns.iter().sum::<f64>() / returns.len() as f64;
        let variance = returns
            .iter()
            .map(|r| (r - avg_return).powi(2))
            .sum::<f64>()
            / returns.len() as f64;
        let std_return = variance.sqrt();

        if std_return.abs() < f64::EPSILON {
            0.0
        } else {
            avg_return / std_return
        }
    }

    /// Calculate how quickly agent recovers from setbacks.
    fn calculate_recovery_speed(&self) -> Option<f64> {
        if self.crisis_decisions.is_empty() {
            return None;
        }

        let mut recovery_speeds = Vec::new();

        for crisis in &self.crisis_decisions {
            let crisis_cycle = crisis.cycle;

            // Look for recovery in next 20 cycles
            for i in (crisis_cycle + 1)..self.decisions.len().min(crisis_cycle + 21) {
                let balance = self.decisions[i]
                    .state
                    .get("balance")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.0);
                let compute = self.decisions[i]
                    .state
                    .get("compute_hours_remaining")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.0);

                if balance > 100.0 && compute > 30.0 {
                    let recovery_time = i - crisis_cycle;
                    recovery_speeds.push(recovery_time as f64);
                    break;
                }
            }
        }

        if recovery_speeds.is_empty() {
            None
        } else {
            let avg_recovery_time =
                recovery_speeds.iter().sum::<f64>() / recovery_speeds.len() as f64;
            // Convert to 0-100 score (20 cycles = 0, 1 cycle = 100)
            Some((100.0 - (avg_recovery_time - 1.0) * 5.0).max(0.0))
        }
    }
}

impl Default for RiskProfiler {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_decision(
        balance: f64,
        compute: f64,
        task_hours: f64,
        company_hours: f64,
    ) -> RiskDecisionRecord {
        let mut state = HashMap::new();
        state.insert("balance".to_string(), serde_json::json!(balance));
        state.insert(
            "compute_hours_remaining".to_string(),
            serde_json::json!(compute),
        );

        let mut action = HashMap::new();
        action.insert("task_work_hours".to_string(), serde_json::json!(task_hours));
        action.insert(
            "company_investment_hours".to_string(),
            serde_json::json!(company_hours),
        );

        RiskDecisionRecord {
            state,
            action,
            reasoning: String::new(),
        }
    }

    #[test]
    fn test_risk_category_from_score() {
        assert_eq!(
            RiskCategory::from_score(10.0),
            RiskCategory::VeryConservative
        );
        assert_eq!(RiskCategory::from_score(30.0), RiskCategory::Conservative);
        assert_eq!(RiskCategory::from_score(50.0), RiskCategory::Moderate);
        assert_eq!(RiskCategory::from_score(70.0), RiskCategory::Aggressive);
        assert_eq!(RiskCategory::from_score(90.0), RiskCategory::VeryAggressive);
    }

    #[test]
    fn test_crisis_severity() {
        assert_eq!(
            CrisisSeverity::from_resources(5.0, 10.0),
            Some(CrisisSeverity::Critical)
        );
        assert_eq!(
            CrisisSeverity::from_resources(15.0, 15.0),
            Some(CrisisSeverity::Severe)
        );
        assert_eq!(
            CrisisSeverity::from_resources(40.0, 25.0),
            Some(CrisisSeverity::Moderate)
        );
        assert_eq!(
            CrisisSeverity::from_resources(80.0, 35.0),
            Some(CrisisSeverity::Mild)
        );
        assert_eq!(CrisisSeverity::from_resources(150.0, 50.0), None);
    }

    #[test]
    fn test_risk_profiler_empty() {
        let profiler = RiskProfiler::new();
        let tolerance = profiler.calculate_risk_tolerance();
        assert_eq!(tolerance.overall_risk_score, 50.0);
        assert_eq!(tolerance.risk_category, RiskCategory::Moderate);
    }

    #[test]
    fn test_crisis_identification() {
        let mut profiler = RiskProfiler::new();
        let decisions = vec![
            make_decision(100.0, 50.0, 4.0, 0.0), // No crisis
            make_decision(15.0, 8.0, 4.0, 0.0),   // Severe
            make_decision(5.0, 3.0, 4.0, 0.0),    // Critical
        ];
        profiler.load_decisions(decisions);

        assert_eq!(profiler.crisis_decisions().len(), 2);
    }

    #[test]
    fn test_conservative_behavior() {
        let mut profiler = RiskProfiler::new();
        let decisions = vec![
            make_decision(30.0, 15.0, 8.0, 0.0), // Low balance, working hard
            make_decision(40.0, 20.0, 6.0, 0.0),
            make_decision(60.0, 30.0, 4.0, 0.0),
        ];
        profiler.load_decisions(decisions);

        let tolerance = profiler.calculate_risk_tolerance();
        assert!(tolerance.overall_risk_score < 50.0);
    }
}
